stdlib/public/Concurrency/AsyncSequence.swift

```diff
@@ -16,5 +16,153 @@ import Swift
 public protocol AsyncSequence {
   associatedtype AsyncIterator: AsyncIteratorProtocol where AsyncIterator.Element == Element
   associatedtype Element
-  func makeAsyncIterator() -> AsyncIterator
+  __consuming func makeAsyncIterator() -> AsyncIterator
+}
+
+extension AsyncSequence {
+  @inlinable
+  public func reduce<Result>(
+    _ initialResult: Result,
+    _ nextPartialResult:
+      (_ partialResult: Result, Element) async throws -> Result
+  ) async rethrows -> Result {
+    var accumulator = initialResult
+    var iterator = makeAsyncIterator()
+    while let element = try await iterator.next() {
+      accumulator = try await nextPartialResult(accumulator, element)
+    }
```

Comment:
Can this be

```
for try await element in self {
  accumulator = try await nextPartialResult(accumulator, element)
}
```

Same for other functions in this file.

Reply:
I am waiting on some stuff Slava was still fixing up for that, it should be in now